torch/optim/svrg.py

```diff
@@ -0,0 +1,174 @@
+                if len(state) == 0:
+                    state['average_gradient'] = torch.zeros_like(p.grad.data)
+                average_gradient = state['average_gradient']
+                snapshot_params = group['snapshot_params'][idx]
+                # gradient data
+                d_p = p.grad.data
+                # subtract the average gradient
+                d_p.add_(average_gradient)
+                # add the snapshot gradient
+                if snapshot_params.grad is not None:
+                    d_p.add_(-1, snapshot_params.grad.data)
+                if weight_decay != 0:
+                    d_p.add_(weight_decay, p.data)
+                if momentum != 0:
+                    param_state = self.state[p]
+                    if 'momentum_buffer' not in param_state:
+                        buf = param_state['momentum_buffer'] = torch.zeros_like(p.data)
+                        buf.mul_(momentum).add_(d_p)
+                    else:
+                        buf = param_state['momentum_buffer']
+                        buf.mul_(momentum).add_(1 - dampening, d_p)
+                    if nesterov:
+                        d_p = d_p.add(momentum, buf)
+                    else:
+                        d_p = buf
+
+                p.data.add_(-group['lr'], d_p)
+
+        return loss
+
+    def update_snapshot(self, dataloader, closure):
+        r"""Updates the parameter snapshot and the average gradient
+
+        Arguments:
+            dataloader : A dataloader used to get the training samples.
+            closure (callable): A closure that reevaluates the snapshot model
+                and returns the loss.
+        """
+        if closure is None:
+            raise RuntimeError("A closure has to be given")
+        if dataloader is None:
+            raise RuntimeError("Dataloader has to be given")
+        # Zero the gradient of the parameters
+        self.zero_grad()
+        # Take a snapshot of the latest parameters
+        for group in self.param_groups:
+            for p, sp in zip(group['params'], group['snapshot_params']):
+                sp.data.copy_(p.data)
+                state = self.state[p]
+                # State initialization, add the average gradient to the state
+                if len(state) == 0:
+                    state['average_gradient'] = torch.zeros_like(p.grad.data)
+        # Iterate over all the dataset to compute the average gradient
+        for i, (data, target) in enumerate(dataloader):
+            closure(data, target)
+            for group in self.param_groups:
+                for idx, p in enumerate(group['snapshot_params']):
+                    state = self.state[p]
+                    if p.grad is None:
+                        continue
+                    if i == 0:
+                        state['average_gradient'].zero_()
+                    state['average_gradient'].add_(1 / len(dataloader), p.grad.data)
```

Comment:
Do we need to take care of adding the L2 reg gradient here (and the snapshot parameters above)? Looks like right now thew weight decay is only applied for the parameters but not for the snapshots.